File: docs/Cours/code/cherche.ml
exception Solution of int

(** cherche tab val vaut Some i où i est l'indice de première
occurence de v dans tab, ou None si v n'apparaît pas dans tab *)
let cherche tab v =
  try for i = 0 to Array.length tab - 1 do
        if tab.(i) = v then raise (Solution i)
      done;
      raise Not_found
  with
  | Solution i -> Some i
  | Not_found -> None

File: docs/Cours/code/cherche.c
/** tableau tab (non NULL) de longueur n,
 * renvoie la distance entre la première et la dernière occurence de val
 * dans tab si val apparaît dans tab, <0 sinon */
int cherche(int *tab, int n, int val){
  int premier, dernier, i;
  assert(tab != NULL && n>0);
  
  for(i=0; i<n; i=i+1){
    if (tab[i] == val) {
      break;
    }
  }
  premier = i;

  for(i=n-1; i>=0; i=i-1){
    if (tab[i] == val) {
      break;
    }
  }
  dernier = i;
  
  return dernier - premier;
}
